### PR TITLE
New module: CONCOCT/CUTUPFASTA

### LINKED ISSUE #2118: new module: CONCOCT/CUTUPFASTA
### Is there an existing module for this?

- [X] I have searched for the existing module

### Is there an open PR for this?

- [X] I have searched for existing PRs

### Is there an open issue for this?

- [X] I have searched for existing issues

### Are you going to work on this?

- [X] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by jfy133**
Adds CONOCT/CUTUPFASTA to close #2118 

## PR checklist

Closes #XXX 

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

File: modules/nf-core/concoct/cutupfasta/main.nf [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        'https://depot.galaxyproject.org/singularity/concoct:1.1.0--py38h7be5676_2':
+        'quay.io/biocontainers/concoct:1.1.0--py38h7be5676_2' }"
+
+    input:
+    tuple val(meta), path(fasta)
+    val(bed)
+
+    output:
+    tuple val(meta), path("*.fasta")                , emit: fasta
+    tuple val(meta), path("*.bed")  , optional: true, emit: bed
```
Comment (matthdsm) [inline modules/nf-core/concoct/cutupfasta/main.nf:17]: Alignment 😱

File: modules/nf-core/concoct/cutupfasta/main.nf [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    path "versions.yml"             , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args       = task.ext.args ?: ''
+    def prefix     = task.ext.prefix ?: "${meta.id}"
+    def bedfile = bed ? "-b ${prefix}.bed" : ""
+    if ("$fasta" == "${prefix}.fasta") error "Input and output names are the same, set prefix in module configuration to disambiguate!"
```
Comment (matthdsm) [inline modules/nf-core/concoct/cutupfasta/main.nf:27]: You could also catch this by using `stageAs` to stage the input in a subdir

File: modules/nf-core/concoct/cutupfasta/main.nf [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    label 'process_single'
+
+    conda (params.enable_conda ? "bioconda::concoct=1.1.0" : null)
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/concoct:1.1.0--py38h7be5676_2':
+        'quay.io/biocontainers/concoct:1.1.0--py38h7be5676_2' }"
+
+    input:
+    tuple val(meta), path(fasta)
+    val(bed)
```
Comment (nvnieuwk) [inline modules/nf-core/concoct/cutupfasta/main.nf:13]: The `bed` can be handled in args, right? `ext.args = {"-b ${meta.id}.bed"}`?

File: modules/nf-core/concoct/cutupfasta/main.nf [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    tuple val(meta), path("*.bed")  , optional: true, emit: bed
+    path "versions.yml"             , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args       = task.ext.args ?: ''
+    def prefix     = task.ext.prefix ?: "${meta.id}"
+    def bedfile = bed ? "-b ${prefix}.bed" : ""
```
Comment (JoseEspinosa) [inline modules/nf-core/concoct/cutupfasta/main.nf:26]:
```suggestion
    def bedfile    = bed ? "-b ${prefix}.bed" : ""
```

File: tests/modules/nf-core/concoct/cutupfasta/main.nf [hunk 5]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env nextflow
+
+nextflow.enable.dsl = 2
+
+include { CONCOCT_CUTUPFASTA                           } from '../../../../../modules/nf-core/concoct/cutupfasta/main.nf'
```
Comment (matthdsm) [inline tests/modules/nf-core/concoct/cutupfasta/main.nf:5]: Too much whitespace?

**Review by nvnieuwk — CHANGES_REQUESTED** [review]

File: modules/nf-core/concoct/cutupfasta/main.nf (hunk 3 above)
Comment (matthdsm) [reply modules/nf-core/concoct/cutupfasta/main.nf:13]: File inputs should be args IMO

File: modules/nf-core/concoct/cutupfasta/meta.yml [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+      type: file
+      description: File containing software versions
+      pattern: "versions.yml"
+  - fasta:
+      type: file
+      description: Cut up fasta file in non-overlapping or overlapping parts of equal length.
+      pattern: "*.fasta"
+  - bed:
+      type: file
+      description: BED File containing locations on original contigs where they were cut up.
```
Comment (JoseEspinosa) [inline modules/nf-core/concoct/cutupfasta/meta.yml:50]:
```suggestion
      description: Optional BED File containing locations on original contigs where they were cut up.
```

File: tests/config/pytest_modules.yml [hunk 7]
```diff
@@ -1,2845 +1,2849 @@
 abacas:
```
Comment (JoseEspinosa) [inline tests/config/pytest_modules.yml:1]: Why are the indents removed? It should be format this way?

File: tests/modules/nf-core/concoct/cutupfasta/main.nf (hunk 5 above)
Comment (JoseEspinosa) [inline tests/modules/nf-core/concoct/cutupfasta/main.nf:5]:
```suggestion
include { CONCOCT_CUTUPFASTA } from '../../../../../modules/nf-core/concoct/cutupfasta/main.nf'
```

File: modules/nf-core/concoct/cutupfasta/main.nf (hunk 3 above)
Comment (nvnieuwk) [reply modules/nf-core/concoct/cutupfasta/main.nf:13]: Yeah good point, but using a random boolean can be restrictive IMO. but yeah nothing stops me from using it in `args` myself while specifying false for the boolean

**Review by JoseEspinosa — APPROVED** [review]
LGTM, just small suggestions and a question about the pytest_modules.yml 🚀 
Feel free to merge after addressing them and when tests are passed 😄

File: modules/nf-core/concoct/cutupfasta/main.nf (hunk 1 above)
Comment (jfy133) [reply modules/nf-core/concoct/cutupfasta/main.nf:17]: :shakesfist:

**Comment by JoseEspinosa** [comment]
> LGTM, just small suggestions and a question about the pytest_modules.yml 🚀 Feel free to merge after addressing them and when tests are passed 😄

And of course addressing the other comments of the reviewing crowd 🤣

File: modules/nf-core/concoct/cutupfasta/main.nf (hunk 3 above)
Comment (matthdsm) [reply modules/nf-core/concoct/cutupfasta/main.nf:13]: I was wrong! It's not a file 😱 
In that case I agree with Nicolas

File: modules/nf-core/concoct/cutupfasta/main.nf (hunk 2 above)
Comment (jfy133) [reply modules/nf-core/concoct/cutupfasta/main.nf:27]: Ooh clever... I'll stick to this now as per guidelines but I might use that trick in the future

**Comment by jfy133** [comment]
@nf-core-bot fix linting

**Pull request merged**